sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -1416,13 +1417,43 @@ case class ValidateExternalType(child: Expression, expected: DataType)
 
   override def nullable: Boolean = child.nullable
 
-  override def dataType: DataType = RowEncoder.externalDataTypeForInput(expected)
-
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported")
+  override val dataType: DataType = RowEncoder.externalDataTypeForInput(expected)
 
   private val errMsg = s" is not a valid external type for schema of ${expected.simpleString}"
 
+
+  private lazy val checkType: (Any) => Boolean = expected match {
+    case _: DecimalType =>
+      (value: Any) => {
+        value.isInstanceOf[java.math.BigDecimal] || value.isInstanceOf[scala.math.BigDecimal] ||
+          value.isInstanceOf[Decimal]
+      }
+    case _: ArrayType =>
+      (value: Any) => {
+        value.getClass.isArray || value.isInstanceOf[Seq[_]]
+      }
+    case _ =>
+      val dataTypeClazz = if (dataType.isInstanceOf[ObjectType]) {
```

Comment:
Ok, two more things.
1. Let's just move this logic into the `classForNativeTypeOf` function.
2. Where are we dealing with `PythonUserDefinedType` or `UserDefinedType`? The `classForNativeTypeOf` function does not appear to be handling those.

Reply:
Both types (`PythonUserDefinedType` or `UserDefinedType`) are converted into native types or ObjectType in `externalDataTypeForInput`. So, IIUC we can handle the two types here, too.